### ISSUE TITLE
Load balancing call - pick subchannel stack in infinite recursion causing app to freeze

### ISSUE THREAD

**Issue by oryaacov** [grpc/grpc-node#2604]
### Problem description
We use @grpc/grpc-js to communicate between our Electron app to a service, communicating by a stream over localhost. Sometimes when the service is restarting and the stream is going down and re-open we stack in open endless recursion, with endless printing of the following log:

```
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Pick called
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Pick called
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.686Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick called
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick called
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick called
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick result: COMPLETE subchannel: (17) ::1:9248 status: undefined undefined
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Picked subchannel (17) ::1:9248 has state IDLE after getting credentials metadata. Retrying pick
D 2023-10-23T21:47:11.687Z | load_balancing_call | [186] Pick called
```

the lines that that keeps re-accruing  
grpc-node/packages/grpc-js/src/load-balancing-call.ts
function doPick line 169.
`              if (
                pickResult.subchannel!.getConnectivityState() !==
                ConnectivityState.READY
              ) {
                this.trace(
                  'Picked subchannel ' +
                    subchannelString +
                    ' has state ' +
                    ConnectivityState[
                      pickResult.subchannel!.getConnectivityState()
                    ] +
                    ' after getting credentials metadata. Retrying pick'
                );
                this.doPick();
                return;`

do pick keeps re-invoking endlessly - causing the application to freeze.

if it help in any way, we are using the following code to connect to the stream:
`async connectToStream(): Promise<void> {
    if (this.subscribeReqId) {
      throw new Error('stream init already in progress');
    }

    await ensureConnectivityState(this.client);

    log.debug('connecting to connect ui stream');

    if (this.stream) {
      this.stream.removeAllListeners();
      this.stream = undefined;
    }

    this.stream = this.buildStream();

    log.debug('subscribeToConnectUiStream');
    this.subscribeToStreamHooks();

    this.subscribeReqId = this.generateRandomRequestId();
    try {
      const res = await new Promise<TRes>((resolve, reject) => {
        const subscriptionRequest = this.buildSubscribeMessage(this.subscribeReqId);
        this.stream.write(subscriptionRequest);
        this.responseCache.set(this.subscribeReqId, {resolve, reject}, this.defualtRequestTimeout);

        log.debug('connected to service stream');
      });

      this.handleStreamSubscriptionResponse(res);
    } finally {
      this.subscribeReqId = undefined;
    }

    this.isStreamConnected = true;
  }`

the line `log.debug('connected to service stream');` is the last one that's been executed before the app freeze and enter the endless recursion loop.
 
### Reproduction steps
Unfortunately that's not easy  to reproduce this happens 1 to every 10-20 that I reset the service killing the stream. I guess that it happens in some weird state of the stream.  

### Environment
 - OS Windows 11 & Mac OS (Monterey)
 - Node version [e.g. v18.12.1]
 - Node installation method nvm on mac, regular setup on windows 
 - Package name and version    "@grpc/grpc-js": "^1.8.10",

**Comment by murgatroid99** [comment]
This looks exactly like the bug reported in #2560. That bug was fixed in version 1.9.2. The latest version of the library is 1.9.7.

**Comment by paulish** [comment]
@murgatroid99 also noticed that the changes in 1.9.7 increased cpu usage by my service from 2% to 100%. Downgrade back to 1.9.5 solved the cpu problem.

**Comment by paulish** [comment]
With 1.9.7/1.9.8 I get the following messages, while with 1.9.5 I don't see them in the logs:
```telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.184Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.185Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 2 -> 3
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | subchannel_refcount | (4) 10.250.200.36:22005 refcount 3 -> 2
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | dns_resolver | Returning IP address for target dns:10.250.200.36:22005
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | pick_first | TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | resolving_load_balancer | dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE
telemedialog_central  | D 2023-10-30T05:34:57.186Z | v1.9.8 1 | connectivity_state | (2) dns:10.250.200.36:22005 TRANSIENT_FAILURE -> TRANSIENT_FAILURE```

**Comment by dixler** [comment]
We're also running into this
- https://github.com/pulumi/pulumi/issues/13687#issuecomment-1783941258

**Comment by murgatroid99** [comment]
I published a change in grpc-js version 1.9.9 that should fix that. Please try it out.

**Comment by paulish** [comment]
@murgatroid99 thanks, the problem has gone with 1.9.9.

**Comment by xTacobaco** [comment]
@murgatroid99 This problem seems to be fixed for us as well with 1.9.9 .
